firedrake/slate/slac.py

```diff
@@ -0,0 +1,512 @@
+
+The Eigen C++ library (http://eigen.tuxfamily.org/) is required, as all low-level numerical
+linear algebra operations are performed using the `Eigen::Matrix` methods built into Eigen.
+"""
+from __future__ import absolute_import, print_function, division
+
+import petsc
+
+from coffee import base as ast
+from coffee.visitor import Visitor
+
+from firedrake.constant import Constant
+from firedrake.tsfc_interface import SplitKernel, KernelInfo, compile_form
+from firedrake.slate.slate import TensorBase, Tensor, UnaryOp, BinaryOp
+from firedrake import op2
+
+from functools import partial
+
+from ufl.algorithms.multifunction import MultiFunction
+from ufl.algorithms.map_integrands import map_integrand_dags
+from ufl.form import Form
+
+from tsfc.parameters import SCALAR_TYPE
+
+
+__all__ = ['compile_slate_expression']
+
+
+class RemoveRestrictions(MultiFunction):
+    """UFL MultiFunction for removing any restrictions on the integrals of forms."""
+
+    expr = MultiFunction.reuse_if_untouched
+
+    def positive_restricted(self, o):
+        return self(o.ufl_operands[0])
+
+
+class Transformer(Visitor):
+    """Replaces all out-put tensor references with a specified
+    name of :type: `Eigen::Matrix` with appropriate shape.
+
+    The default name of :data:`"A"` is assigned, otherwise a
+    specified name may be passed as the :data:`name` keyword
+    argument when calling the visitor."""
+
+    def visit_object(self, o, *args, **kwargs):
+        """Visits an object and returns it.
+
+        e.g. string ---> string
+        """
+        return o
+
+    def visit_list(self, o, *args, **kwargs):
+        """Visits an input of COFFEE objects and returns
+        the complete list of said objects."""
+
+        newlist = [self.visit(e, *args, **kwargs) for e in o]
+        if all(newo is e for newo, e in zip(newlist, o)):
+            return o
+
+        return newlist
+
+    visit_Node = Visitor.maybe_reconstruct
+
+    def visit_FunDecl(self, o, *args, **kwargs):
+        """Visits a COFFEE FunDecl object and reconstructs
+        the FunDecl body and header to generate
+        Eigen::MatrixBase C++ template functions.
+
+        Creates a template function for each subkernel form
+        template <typename Derived>:
+
+        template <typename Derived>
+        static inline void foo(Eigen::MatrixBase<Derived> const & A, ...)
+        {
+          [Body...]
+        }
+        """
+
+        name = kwargs.get("name", "A")
+        new = self.visit_Node(o, *args, **kwargs)
+        ops, okwargs = new.operands()
+        if all(new is old for new, old in zip(ops, o.operands()[0])):
+            return o
+
+        template = "template <typename Derived>"
+        body = ops[3]
+        args, _ = body.operands()
+        eigen_arg = "Eigen::MatrixBase<Derived> & {0} = const_cast<Eigen::MatrixBase<Derived> &>({0}_);\n".format(name)
+        nargs = [ast.FlatBlock(eigen_arg)] + args
+        ops[3] = nargs
+        ops[6] = template
```

Comment:
It happens that in this case, the list ops is newly consed every time you call operands, so you're not underhandedly modifying something in place.  However, this is unsafe.  Better to shallow-copy the list and reconstruct it.  Note to @FabioLuporini, we should change coffee operands to return tuples, not lists, so that this is not possible.